Lab 5 - movie explore app

princoo opened the pull request:
### ✨ Features Implemented:

* **Component-Based Architecture**: Created reusable React components for layout, movie cards, search, filters.
* **Routing with React Router**: Integrated client-side routing for  navigation across Home, Movie Details, and watchlist pages.
* **State Management with Redux**: Used Redux to manage global state, including movie data, loading states, querryParams,watchlist and reviews.
* **API Integration with Axios**: Fetched movie data from an external API using Axios with  error handling and async action dispatching.
* **Firebase**: configured firebase to store the reviews for specific movies.

* **Testing Setup**:
  * Configured **Jest** and **React Testing Library (RTL)**.
  * Added unit and integration tests for key components and Redux logic.

princoo committed:
chore:

- set up the environment (ts,tailwind,vite)
- create custom components for reusability
- configure Axios instance
- create custom hook for regural fetches

princoo committed:
feat: integrate react-hook-form for search functionality and enhance context management

Chartine02 commented on movieExploreApp/src/components/MovieCard.tsx:
```diff
@@ -0,0 +1,46 @@
+                ? movie.primaryImage.url
+                : "https://www.behance.net/gallery/6312333/File-Not-Found-Movie-Poster/modules/207739455"
+            }
+            alt={movie.primaryImage?.id}
```
The `alt` is meant to be an alternative text of the image.
I would recommend to use the movie's title instead of the image's id.

Chartine02 commented on movieExploreApp/src/components/SideBar.tsx:
```diff
@@ -0,0 +1,94 @@
+  function handleClick(value: string) {
+    setParams((prev) => ({ ...prev, genre: value }));
+  }
+  const genreList: string[] = [
```
It's important to fetch the genreList from the API to ensure it's up to date.
But if you're choosing not to because the genres rarely change, that's a valid approach too. In that case, I’d recommend creating a dedicated `data `folder and storing the genre list there separately.

Chartine02 commented on movieExploreApp/src/components/SideBar.tsx:
```diff
@@ -0,0 +1,94 @@
+        className="inline-flex items-center p-2 mt-2 ms-3 text-sm text-gray-500 rounded-lg sm:hidden hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-200 dark:text-gray-400 dark:hover:bg-gray-700 dark:focus:ring-gray-600"
+      >
+        <span className="sr-only">Open sidebar</span>
+        <svg
```
I would suggest storing the images and icons in `assets`

Chartine02 commented on movieExploreApp/src/main.tsx:
```diff
@@ -0,0 +1,10 @@
+import App from './App.tsx'
+
+createRoot(document.getElementById('root')!).render(
+  //<StrictMode>
```
Is there any specific reasong why you commented `StrictMode` component ?

Chartine02 commented on movieExploreApp/src/types/testData.ts:
I would suggest to store `testData `in `data` as well.

Chartine02 commented on movieExploreApp/src/hooks/useFetch.ts:
```diff
@@ -0,0 +1,36 @@
+  const fetchData = useCallback(async () => {
+    setLoading(true);
+    setError(null);
+    await api
+      .get(url, { params })
+      .then((res) => {
+        setData(res.data.results);
+      })
+      .catch((err: AxiosError) => {
+        setError(err.message);
+      })
+      .finally(() => setLoading(false));
```
I recommend choosing either the `async/await` pattern or the `.then()/.catch()` promise chain and sticking with it for consistency.

- `await `uses `try/catch`
- `.then()/.catch()` uses promise chains

Chartine02 commented on movieExploreApp/.env:
It is always important to include the `.env` file in your gitignore.
You can consider adding it.

Chartine02 reviewed (commented):
Great work with your progress so far!
I have left some reviews.

princoo committed:
feat: Add new components and functionality for movie explore

princoo committed:
fix: enable StrictMode for improved performance and debugging

princoo committed:
feat: update movie data structure and enhance movie details display

princoo committed:
fix: enable StrictMode for improved performance and debugging

princoo committed:
feat: add GitHub Actions workflow for deployment to GitHub Pages

princoo committed:
refactor: moving the project to the root directory

princoo committed:
fix: correct branch name in deployment workflow

princoo committed:
refactor: remove unused React import from StarIcon component and update dummyMovies data structure with ratings and genres

princoo committed:
fix: update base path in Vite configuration for correct deployment

princoo committed:
fix: update base path in Vite config

princoo committed:
fix: correct destination directory name in deployment workflow

princoo committed:
fix: correct destination directory name in deployment workflow

princoo committed:
fix: correct destination directory name in deployment workflow

princoo committed:
fix: correct spelling of 'Weekly' in Vite config base path

princoo committed:
fix: remove extra space in sidebar button className

princoo committed:
fix: update default URL in Hero component and clean up unused code in useSearch hook

princoo committed:
fix: add environment variables for build step in deployment workflow

princoo committed:
fix: format code for better readability and consistency across components

princoo committed:
fix:

- update context parameters handling in Hero and SideBar components
- adjust selected movie index in MoviesContainer, and clean up SearchBar component

Chartine02 commented on README.md:
Update the `README` file to descripte the current project.

Chartine02 commented on src/Pages/Hero.tsx:
```diff
@@ -0,0 +1,111 @@
+export default function Hero() {
+  const contextParams = useContext<MyContextType>(ParamContext);
+  const [url, setUrl] = useState<string>("titles");
+  const [seachValue, setseachValue] = useState<string>("");
```
I would sugges to fix the typo and use pascal casing consinstently
`const [searchValue, setSearchValue] = useState<string>("");`

princoo replied:
for variables and functions, I will stick to camelCase to follow standard conventions. If I’ve accidentally used PascalCase somewhere it shouldn't be, I willl go through and update it. Appreciate the catch!

Chartine02 commented on src/components/MovieCard.tsx:
```diff
@@ -0,0 +1,43 @@
+  onClick,
+}: {
+  movie: Movie;
+  onClick: (data: Movie) => void;
```
I would suggest extracting the the interface and define it outside the function

Chartine02 commented on src/components/MovieCard.tsx:
```diff
@@ -0,0 +1,43 @@
+              loaded ? "opacity-100" : "opacity-0 blur-sm"
+            }`}
+            src={movie.primaryImage ? movie.primaryImage.url : defaultImage}
+            alt={movie.primaryImage?.id}
```
The `alt` provides more information about the image,so I would suggest using the title of the movie instead.

Chartine02 commented on src/components/MovieDetails.tsx:
```diff
@@ -0,0 +1,48 @@
+import { useState } from "react";
+
+export default function MovieDetails({ movie }: { movie: Movie }) {
+  const [loaded, setLoaded] = useState(false);
```
You're creating and updating this state across multiple files. I suggest finding a way to centralize the logic so you don't have to repeat the same implementation in each place

Chartine02 commented on src/components/Footer.tsx:
```diff
@@ -0,0 +1,11 @@
+export default function Footer() {
+  return (
+    <footer className="pt-10">
+        <div className="mt-10  items-center justify-center sm:flex">
```
Consider using `py` to add padding to the footer element instead adding pt both elements.

Chartine02 commented on src/components/Paginator.tsx:
```diff
@@ -0,0 +1,47 @@
+  onNext,
+  page,
+}: {
+  onNext: () => void;
```
You can consider extracting the interface outside of the function

princoo committed:
fix: update README content, add image asset, and refactor context usage in components

princoo committed:
feat: add ImageWithLoading component for improved image loading experience in MovieCard and MovieDetails

princoo committed:
feat: implement Redux slices for movies and watchlist management, add Redux Toolkit dependencies

princoo committed:
feat: add routing and movie page components, integrate Redux for state management, and update environment configurations

princoo committed:
refactor: clean up unused imports and comments across components, improve MovieCard and MoviesContainer functionality

princoo committed:
feat: integrate react-hot-toast for notifications, implement watchlist functionality, and enhance routing with new components

princoo committed:
chore: update .gitignore to include environment configuration files

princoo committed:
chore: remove environment configuration files for development and production

princoo committed:
refactor: reorganize router configuration for improved readability and maintainability

princoo committed:
refactor: remove unused loadEnv import and clean up commented code in vite.config.ts

princoo committed:
feat: integrate Firebase for reviews functionality

- Added Firebase configuration and services for storing and retrieving movie reviews.
- Implemented review submission and display components including ReviewForm, ReviewList, and ReviewCard.
- Enhanced MoviePage to include review functionality with a toggle for the review form.
- Updated Redux store to manage reviews state with a new reviewsSlice.
- Added utility function for formatting review timestamps.
- Introduced a new InputField component for form inputs.
- Updated package.json to include Firebase SDK.
- Improved UI components with new styles and responsive design.

princoo committed:
feat(tests):

- implement tests for reusable components

princoo committed:
feat(theme):

- setting up theme
- toogle between light and dark mode
- persits the theme preference

Bob-Karemera-Shema reviewed (approved):
I am impressed by the quality of the work. You introduced a robust set of features for a movie exploration app, with well-structured React components, Redux for state management, React Router for navigation, and external integrations like Axios and Firebase. The setup for testing using Jest and RTL is also commendable. The architecture is scalable and adheres to modern best practices. Given the large number of changed files, 89 in total, I suggested that future PRs be broken into smaller chunks to make them easier to review. Overall, this is a good PR.
